Add explicit renameColumn method for Table

Tofandel opened the pull request:
<!-- Fill in the relevant information below to help triage your pull request. -->

|      Q       |   A
|------------- | -----------
| Type         | feature
| Fixed issues | https://github.com/doctrine/migrations/issues/57

#### Summary

<!-- Provide a summary of your change. -->
Reopening #6078  on 3.7.x

Add a `renameColumn` in the `Table` class

This allow to create migration with schemas that rename columns explicitely, which is a missing feature

Right now only implicit detection is done (it looks for added columns with the same definition as a dropped column), and that means if you change an option or the type of the column, it will not consider it as a rename anymore and instead will drop and recreate, by being explicit, this is not a problem anymore


This PR allows the following migration, without losing the column data
```php

    public function up(Schema $schema): void
    {
        $schema->getTable('test')->renameColumn('foo', 'bar')->setType(Type::getType('text'));
    }
    
    public function down(Schema $schema): void
    {
        $schema->getTable('test')->renameColumn('bar', 'foo')->setType(Type::getType('string'));
    }
 ```
 
 Which before would have to drop the column and recreate it, as well as needing a full redefinition of the column, and losing the data in the process
 
```php

    public function up(Schema $schema): void
    {
        $schema->getTable('test')->dropColumn('foo')->addColumn('bar', 'text');
    }
    
    public function down(Schema $schema): void
    {
        $schema->getTable('test')->dropColumn('bar')->addColumn('foo', 'string');
    }
 ```
 
 PS: I'm building a feature for the `doctrine/migration` package to generate diffs of schema, directly using `Schema` instead of platform specific sql and without this `renameColumn` this would be impossible, it would also allow other frameworks which use schemas (Yes I'm thinking laravel) to rename columns [without using internals](https://github.com/laravel/framework/blob/10.x/src/Illuminate/Database/Schema/Grammars/RenameColumn.php#L65C21-L65C35)

Tofandel committed:
Add renameColumn() method on Table

This allow to create migrations with schemas that rename columns
explicitly, as opposed to dropping one column and creating another one.

Co-authored-by: Grégoire Paris <postmaster@greg0ire.fr>

Tofandel commented on tests/Platforms/DB2PlatformTest.php:
```diff
@@ -128,7 +128,8 @@ public function getAlterTableColumnCommentsSQL(): array
         return [
             'ALTER TABLE mytable ' .
-            'ADD COLUMN quota INTEGER NOT NULL WITH DEFAULT',
+            'ADD COLUMN quota INTEGER NOT NULL WITH DEFAULT ' .
```
See https://github.com/doctrine/dbal/blob/0febc80c975436f351792a254e96e0e7af3f72b3/tests/Platforms/AbstractPlatformTestCase.php#L432-L440
To understand why those tests were incorrect

The column is renamed in the test, but not in the test case output

Tofandel commented:
This PR can be followed by a deprecation and removal of the `renamedColumns` in the `TableDiff` as this is now redundant and can be put into `modifiedColumns`

greg0ire commented on src/Schema/Table.php:
```diff
@@ -26,6 +26,9 @@ class Table extends AbstractAsset
     /** @var Column[] */
     protected $_columns = [];
 
+    /** @var string[] */
```
```suggestion
    /** @var array<string, string> keys are new names, values are old names */
```

greg0ire commented on src/Schema/Table.php:
```diff
@@ -26,6 +26,9 @@ class Table extends AbstractAsset
     protected $_columns = [];
 
+    /** @var string[] */
+    protected array $_renamedColumns = [];
```
I would give as little access as possible to this. We can always loosen access afterwards if required. Also, let's loose the underscore.

```suggestion
    private array $renamedColumns = [];
```

greg0ire commented on src/Schema/Table.php:
```diff
@@ -346,6 +349,34 @@ public function addColumn($name, $typeName, array $options = [])
         return $column;
     }
 
+    /** @return string[] */
```
```suggestion
    /** @return array<string, string> */
```

greg0ire commented on src/Schema/Table.php:
```diff
@@ -346,6 +349,34 @@ public function addColumn($name, $typeName, array $options = [])
     }
 
+    /** @return string[] */
+    public function getRenamedColumns(): array
```
```suggestion
    final public function getRenamedColumns(): array
```

greg0ire commented on src/Schema/Table.php:
```diff
@@ -346,6 +349,34 @@ public function addColumn($name, $typeName, array $options = [])
+        return $this->_renamedColumns;
+    }
+
+    public function renameColumn(string $oldName, string $newName): Column
```
```suggestion
    final public function renameColumn(string $oldName, string $newName): Column
```

greg0ire commented on src/Platforms/DB2Platform.php:
```diff
@@ -634,6 +635,17 @@ public function getAlterTableSQL(TableDiff $diff)
+
+            $oldColumnName = $oldColumn->getQuotedName($this);
+
+            // Column names in DB2 are case insensitive and automatically uppercased on the server.
```
So… if a user tries to change the case, we ignore them silently? Shouldn't we instead let them know there is no point in doing this?

Tofandel replied:
If we do a check, it should likely be during the `renameColumn` use directly, because from what I gathered all the availlable database drivers are case insensitive, and the problem I had was when a column is created with a lowercase column (eg: `c1`), some drivers (like Oracle and SQL server) actually creates it all uppercase (eg: `C1`) and then when computing the diff the `oldColumnName` is uppercase while the current column name is lowercase even though the user didn't actually request for a change but it was computed automatically which would trigger an unnecessary error

It's a broader issue than this PR can tackle but I do think the column name normalization should maybe take care of this, or maybe when during the diff computing beforehand

We could for instance add a property to the driver to determine if the table name is case sensitive or not, and same for columns and then normalize based on the flags

Tofandel commented on tests/Functional/Platform/RenameColumnTest.php:
```diff
@@ -33,6 +35,35 @@ public function testColumnPositionRetainedAfterRenaming(string $columnName, stri
+        $this->dropAndCreateTable($table);
+
+        // Force a different type to make sure it's not being caught implicitly
+        $table->renameColumn($columnName, $newColumnName)->setType(Type::getType(Types::BIGINT))->setLength(32);
```
It's worth noting that some drivers will fail to change some types to incompatible types, it doesn't have anything to do with this PR though, it's just natively changing the type of a column to another specific type which is sometimes not supported, the biggest offender being DB2 which is unable to transform a `string`  type to a `text` type (which I think might be a common change), it's impossible to know all of this in advance and unlikely we could cover all edge cases, so I'm leaving it to the driver to throw an error when the query fails, as was already the case

It's not a blocker for this PR and just a possible future improvement to the offending drivers if the need really arises as we can still test and implement the most common type changes

Tofandel commented:
Just bumping this so it doesn't get forgotten

Is there anything that should be changed or worth adding to this PR after my comments? Do I need another review?

derrabus commented:
Sorry, I simply haven't had time to take another look at this PR. I'm adding it to the milestone so I hopefully won't forget it.

derrabus commented on tests/Functional/Schema/SchemaManagerFunctionalTestCase.php:
```diff
@@ -1377,10 +1377,10 @@ public function testAlterColumnComment(
-        $onlineTable->changeColumn('comment2', ['comment' => $comment1]);
-        $onlineTable->changeColumn('no_comment1', ['comment' => $comment1]);
-        $onlineTable->changeColumn('no_comment2', ['comment' => $comment2]);
+        $onlineTable->modifyColumn('comment1', ['comment' => $comment2]);
+        $onlineTable->modifyColumn('comment2', ['comment' => $comment1]);
+        $onlineTable->modifyColumn('no_comment1', ['comment' => $comment1]);
+        $onlineTable->modifyColumn('no_comment2', ['comment' => $comment2]);
```
This looks like an unrelated change?

Tofandel replied:
Yes it was

derrabus commented on src/Platforms/SQLServerPlatform.php:
```diff
@@ -560,12 +561,25 @@ public function getAlterTableSQL(TableDiff $diff)
+
+            // Column names in SQL server are case insensitive and automatically uppercased on the server.
+            if (strcasecmp($oldColumnName, $newColumnName) !== 0) {
+                $sql[] = "sp_rename '" . $tableNameSQL . '.' . $oldColumnName .
```
```suggestion
                $sql[] = "EXEC sp_rename '" . $tableNameSQL . '.' . $oldColumnName .
```

Tofandel commented:
I've been cleaning up the PR, and in the end decided to do some deprecations directly in this PR so it's more clear

I deprecated the `getRenamedColums` and for backwards compatibility I changed `getModifiedColumns` to `getChangedColumns` so that `getRenamedColum` and `getModifiedColumns` will still output exactly the same as before, but `getChangedColumns` will be a mix of the renamed and modified columns

I also digged a bit on the case sensitivty issue, none of them are case sensitive (unless the identifier is quoted) and the `Comparator` is already case insensitive (strtolower in Table::normalizeIdentifier), so we are already assuming everyone is using lower case column names, so for now I don't think it's necessary to handle case sensitive renames at all, so I'm just throwing an exception if we use `renameColumn('foo', 'Foo')`

Unrelated: Does anyone have the issue where phpunit just exits when it encounters a skipped test? I've been trying to run them but it stops in the middle because of this I can't run all the tests locally

Tofandel commented on tests/Platforms/SQLServerPlatformTest.php:
```diff
@@ -1427,7 +1430,6 @@ public static function getGeneratesIdentifierNamesInAlterTableSQL(): iterable
                     "CONSTRAINT DF_6B2BD609_4AD86123 DEFAULT 'foo'",
                     'ALTER TABLE mytable DROP COLUMN removecolumn',
                     'ALTER TABLE mytable DROP CONSTRAINT DF_6B2BD609_9BADD926',
-                    'ALTER TABLE mytable ALTER COLUMN mycolumn NVARCHAR(255) NOT NULL',
```
This statement was not needed, if you look at the column diff

```php
[
    'mycolumn' => new ColumnDiff(
        'mycolumn',
        new Column('mycolumn', Type::getType(Types::STRING), ['default' => 'bar']),
        ['default'],
        new Column('mycolumn', Type::getType(Types::STRING), ['default' => 'foo']),
    ),
],
```

Because only the `default` value changed, the type didn't

Tofandel commented on tests/Platforms/SQLServerPlatformTest.php:
```diff
@@ -1022,7 +1025,7 @@ public function testGeneratesAlterTableSQLWithColumnComments(): void
                 'ALTER TABLE mytable ALTER COLUMN [create] VARCHAR(MAX) NOT NULL',
                 'ALTER TABLE mytable ALTER COLUMN commented_type INT NOT NULL',
-
+                'ALTER TABLE mytable ALTER COLUMN comment_with_string_literal_char NVARCHAR(255) NOT NULL',
```
The type of the changed columns differ, but `type` is missing from the `changedProperties` parameter

```php
$tableDiff->changedColumns['comment_with_string_literal_char'] = new ColumnDiff(
    'comment_with_string_literal_char',
    new Column('comment_with_string_literal_char', Type::getType(Types::STRING), ['comment' => "'"]),
    ['comment'], // Here type is missing
    new Column('comment_with_string_literal_char', Type::getType(Types::ARRAY), ['comment' => "O'Reilly"]),
);
```
It might be worth reviewing this test, do we keep the type change? Or do I keep the previous behavior if the type is changed but is not present in the `changedProperties`?

Seeing, that the column was first declared as `Types::ARRAY` I do believe it was just a missing `type` in the `changedProperties` list and so this line should stay

Tofandel commented on src/Schema/TableDiff.php:
```diff
@@ -238,10 +248,34 @@ public function getAddedColumns(): array
+    /** @return list<ColumnDiff> */
+    public function getChangedColumns(): array
+    {
+        $legacyChangedColumns = [];
+        foreach ($this->renamedColumns as $oldName => $column) {
+            $legacyChangedColumns[$oldName] = new ColumnDiff($oldName, $column, []);
+        }
```
I had to include backwards compatibility for packages modifying the internals like laravel because they are public, even though marked internal (I see in 4.x they are private, so it's good this PR is here or they would have no alternative)
https://github.com/laravel/framework/blob/10.x/src/Illuminate/Database/Schema/Grammars/RenameColumn.php#L64-L67

stof commented:
The automatic detection of renames should be kept IMO, as not all Schema diffing happens for a Schema that has been manipulating based on the introspected one.
The common case for doctrine/migrations is that the introspected schema is compared to an ideal schema built based on the ORM mapping.

> PS: I'm building a feature for the `doctrine/migration` package to generate diffs of schema, directly using `Schema` instead of platform specific sql and without this `renameColumn` this would be impossible, it would also allow other frameworks which use schemas (Yes I'm thinking laravel) to rename columns [without using internals](https://github.com/laravel/framework/blob/10.x/src/Illuminate/Database/Schema/Grammars/RenameColumn.php#L65C21-L65C35)

Note that Schema-based migrations have 2 big drawbacks compared to generating the SQL at generation time:

- they are much slower as they require introspecting the database between each executed migration (they cannot rely on using the schema altered in the previous migration as extra SQL queries might also have been executed to do more changes)
- they are unable to migrate data, making them safe to run only on an empty database (when generating SQL, you can add extra queries in the middle to perform the necessary data updates).

stof commented on src/Schema/TableDiff.php:
```diff
@@ -238,10 +251,44 @@ public function getAddedColumns(): array
+            $legacyChangedColumns[$oldName] = new ColumnDiff($oldName, $column, []);
+        }
+
+        return array_values(array_merge($this->changedColumns, $legacyChangedColumns));
```
what if we already had a ColumnDiff for that column ?

Tofandel replied:
It's unlikely to happen as that would mean the user is modifying the `renamedColumn` internal property directly from an existing diff which also has changed columns passed in parameters (now all the dbal uses `changedColumns` only, I could handle the case and merge the diff though in the BC code I added a bit higher

Tofandel commented:
> The automatic detection of renames should be kept IMO, as not all Schema diffing happens for a Schema that has been manipulating based on the introspected one.

It has been kept, it's just the internal handling of where the renamed columns are stored that is changed

Tofandel commented on src/Schema/Comparator.php:
```diff
@@ -363,7 +363,29 @@ public function compareTables(Table $fromTable, Table $toTable): TableDiff
+            );
+        }
+
+        $this->detectRenamedColumns($modifiedColumns, $addedColumns, $droppedColumns);
```
@stof 
> The automatic detection of renames should be kept IMO, as not all Schema diffing happens for a Schema that has been manipulating based on the introspected one.

It has been kept, it's just the internal handling of where the renamed columns are stored that is changed, see here, the detected columns are just stored in `modifiedColumns` because that's what the platforms now uses to run the rename queries

Tofandel commented:
> Note that Schema-based migrations have 2 big drawbacks compared to generating the SQL at generation time:
> 
> they are much slower as they require introspecting the database between each executed migration (they cannot rely on using the schema altered in the previous migration as extra SQL queries might also have been executed to do more changes)
> they are unable to migrate data, making them safe to run only on an empty database (when generating SQL, you can add extra queries in the middle to perform the necessary data updates).

Thanks I'll keep that in mind, I don't think for migration the performance drawback of introspecting the table is a big concern, they just run once after all, and usually there is only one to run (if generated with `doctrine:migrations:diff`), so we're only talking milliseconds on deployment commands

For the migration, well now they would be able to rename columns, so it's still safe to run on non empty database to drop, add or rename columns, but yes for anything more advanced, the user will have to use DQL to make queries to migrate data, the idea is just to generate the diff schema for the user to review but as a schema instead of queries, I'm basically doing what `doctrine:migrations:diff` does but outputting php with schema calls, it's just a question of being able to help in making portable apps

stof commented:
@Tofandel A common setup for migrations is that you run them from the start to build new environments (when doing a setup in local environments or in CI). They are only a one-time run __per database__.
In the past, we optimized the case of not using the Schema in the migration (by making the schema introspection lazy thanks to a proxy) and this resulted in dividing the execution time of `doctrine:migration:migrate` by a factor 100 for such setup migrating many versions from scratch. 

> > Note that Schema-based migrations have 2 big drawbacks compared to generating the SQL at generation time:
> > they are much slower as they require introspecting the database between each executed migration (they cannot rely on using the schema altered in the previous migration as extra SQL queries might also have been executed to do more changes)
> > they are unable to migrate data, making them safe to run only on an empty database (when generating SQL, you can add extra queries in the middle to perform the necessary data updates).
> 
> Thanks I'll keep that in mind, I don't think for migration the performance drawback of introspecting the table is a big concern, they just run once after all, and usually there is only one to run (if generated with `doctrine:migrations:diff`), so we're only talking milliseconds on deployment commands
> 
> For the migration, well now they would be able to rename columns, so it's still safe to run on non empty database to drop, add or rename columns, but yes for anything more advanced, the user will have to use DQL to make queries to migrate data, the idea is just to generate the diff schema for the user to review but as a schema instead of queries, I'm basically doing what `doctrine:migrations:diff` does but outputting php with schema calls, it's just a question of being able to help in making portable apps

That's precisely my point: migrating data often needs to happen in the **middle** of the schema change (after introducing new fields but before deleting old ones). If your migration class performs changes on the schema and lets the Migrator determine the queries for the schema diff, you cannot put queries in the middle (you even cannot put queries _after_ as any query added with `addSql` in the `up`/`down` methods would end up before the schema diff queries in the list of queries).

The big difference with what `doctrine:migrations:diff` does is that you can edit the class after generating it.

Tofandel commented:
> The big difference with what doctrine:migrations:diff does is that you can edit the class after generating it.

That's also the case with the command I'm making, it generates a migration but manipulating the schema parameter directly, if you need to separate it into multiple migrations it's possible, after all it just generates migration files for you and you are free to edit them after that

> If your migration class performs changes on the schema and lets the Migrator determine the queries for the schema diff, you cannot put queries in the middle (you even cannot put queries after as any query added with addSql in the up/down methods would end up before the schema diff queries in the list of queries).

To solve this, I'm thinking we can add a feature to add DQL queries to execute in the schema builder directly, if we can indicates queries should run before the schema changes, or after columns are modified but before column deletion or entirely after, then we can completely cut the need to split the schema in multiple migrations, if that's something you agree makes sense, I can investigate the doability when I have time

> In the past, we optimized the case of not using the Schema in the migration (by making the schema introspection lazy thanks to a proxy) and this resulted in dividing the execution time of doctrine:migration:migrate by a factor 100 for such setup migrating many versions from scratch.

Is it not the case anymore?

stof commented:
> > The big difference with what doctrine:migrations:diff does is that you can edit the class after generating it.
> 
> That's also the case with the command I'm making, if you need to separate it into multiple migrations it's possible, after all it just generates migration files for you
> 
> > If your migration class performs changes on the schema and lets the Migrator determine the queries for the schema diff, you cannot put queries in the middle (you even cannot put queries after as any query added with addSql in the up/down methods would end up before the schema diff queries in the list of queries).
> 
> To solve this, I'm thinking we can add a feature to add DQL queries to execute in the schema builder directly, if we can indicates queries should run before the schema changes, or after columns are modified but before column deletion or entirely after, then we can completely cut the need to split the schema in multiple migrations

It will generate a migration file that performs changes on the Schema. And that's precisely my point: you can edit the migration file, but you cannot edit the list of executed queries (as the queries related to the schema comparison will be determined during the execution of the migration, and will be added after any query added in `up`

> Is it not the case anymore?

What are you referring to here ?

Tofandel commented:
> What are you referring to here ?

The optimisation on the introspection

> you can edit the migration file, but you cannot edit the list of executed queries (as the queries related to the schema comparison will be determined during the execution of the migration, and will be added after any query added in up

I don't see why would you need to edit the queries being run, its supposed to be just about platform independant schema modification, no raw SQL allowed, if you use raw SQL, then there is no point in using the schema-diff command

For reference this is what the command generates
```php
    public function up(Schema $schema): void
    {
        // this up() migration is auto-generated, please modify it to your needs

        $case_file = $schema->createTable('case_file')
            ->addOption('create_options', [])
            ->addOption('collation', 'utf8mb4_unicode_ci')
            ->addOption('charset', 'utf8mb4');
        $case_file->addColumn('id', 'ulid');
     }
     
    public function down(Schema $schema): void
    {
        // this down() migration is auto-generated, please modify it to your needs
        $schema->dropTable('case_file');
  }
```

If you need to migrate data, I'm proposing above to add DQL queries in the schema builder for this exact case

stof commented:
@Tofandel I don't see how adding DQL queries (or most like SQL ones as DBAL does not know about DQL at all) in the schema builder would work.
The SQL queries don't make sense in a Schema. They might make sense in a SchemaDiff (but even then, I fail to see how you would represent that to be able to inject those queries in the middle of the list of queries returned by the platform for one operation of the schema diff).

Side note: in some cases, it even requires editing the DDL statements and not just adding extra queries. For instance, in Postgresql (maybe also in other platforms), adding a non-nullable field without default value works only if the table is empty. For a non-empty table, you need to add the field without the `NOT NULL` clause, then populate the field and then alter it to add `NOT NULL`.

Tofandel commented:
> I don't see how adding DQL queries (or most like SQL ones as DBAL does not know about DQL at all) in the schema builder would work.
> The SQL queries don't make sense in a Schema. They might make sense in a SchemaDiff (but even then, I fail to see how you would represent that to be able to inject those queries in the middle of the list of queries returned by the platform for one operation of the schema diff).

I'm thinking more of 3 callbacks in the Schema, up to you to run whatever you want, so we don't need to know about DQL, but yes it does make more sense in a SchemaDiff, but it's not what we're working with in a migration

So maybe if `Schema` is not the place then we can add another `MigrationEvent` parameter to the migration's `up` and `down` this parameter would allow to add the callbacks where they are needed, I didn't yet investigate this that's why I'm not sure yet how it should/could be implemented, but there is definitely possible solutions I'm thinking about

> Side note: in some cases, it even requires editing the DDL statements and not just adding extra queries. For instance, in Postgresql (maybe also in other platforms), adding a non-nullable field without default value works only if the table is empty. For a non-empty table, you need to add the field without the NOT NULL clause, then populate the field and then alter it to add NOT NULL.

It's the same with mysql and mariadb and pretty much all platforms, it makes sense though, if you create a non nullable field without a default, with existing data it doesn't make much sense, because it would normally put them null, but you don't allow null, so what else can it do except error out? You need to create it with a default value, and then change it to be without a default value in this case

stof commented:
> > I don't see how adding DQL queries (or most like SQL ones as DBAL does not know about DQL at all) in the schema builder would work.
> > The SQL queries don't make sense in a Schema. They might make sense in a SchemaDiff (but even then, I fail to see how you would represent that to be able to inject those queries in the middle of the list of queries returned by the platform for one operation of the schema diff).
> 
> I'm thinking more of 3 callbacks in the Schema, up to you to run whatever you want, so we don't need to know about DQL, but yes it does make more sense in a SchemaDiff, but it's not what we're working with in a migration
> 
> So maybe if `Schema` is not the place then we can add another `MigrationEvent` parameter to the migration's `up` and `down` this parameter would allow to add the callbacks where they are needed, I didn't yet investigate this that's why I'm not sure yet how it should/could be implemented, but there is definitely possible solutions I'm thinking about
> 
> > Side note: in some cases, it even requires editing the DDL statements and not just adding extra queries. For instance, in Postgresql (maybe also in other platforms), adding a non-nullable field without default value works only if the table is empty. For a non-empty table, you need to add the field without the NOT NULL clause, then populate the field and then alter it to add NOT NULL.
> 
> It's the same with mysql and mariadb and pretty much all platforms, it makes sense though, if you create a non nullable field without a default, with existing data it doesn't make much sense, because it would normally put them null, but you don't allow null, so what else can it do except error out? You need to create it with a default value, and then change it to be without a default value in this case

and that's a case where the only way to do that is to manually edit the schema diff queries, which you cannot do if your migration only edits a Schema object as it does not have access to customizing the queries.

Tofandel commented:
Well from my point of view you would just create 2 migrations, one with the default value, and one to change the schema again without the default value.

Using the `Schema` is just a different way to work with migrations, not everyone is gonna use them, but it's already possible to write migrations using the schema, though there is no command or utils to generate them automatically currently, the tools we do have only generate raw SQL migrations, so I just think it's a pain point we can alleviate for developers who want to make their apps platform independant

I also think It's also quite a good tool for the first migration in general, it's not going to be bulletproof, just like how the current schema diff is not bulletproof, but it's a start

derrabus commented:
I feel like we should not postpone 3.7.0 any longer, so I'm moving this to 3.8. Sorry that I haven't had the time yet to help getting this PR ready in time. 😓

derrabus commented:
I'm reviewing the PR now. I'm terribly sorry that I've postponed it for so long. One thing though: Codecov complains that we've lost coverage for the old API. Can you check that please?

derrabus commented on src/Schema/Table.php:
```diff
@@ -346,6 +349,44 @@ public function addColumn($name, $typeName, array $options = [])
+        $newName = $this->normalizeIdentifier($newName);
+
+        if ($oldName === $newName) {
+            throw new Exception("You are trying to rename the column ${oldName} but you didn't change it's name");
```
```suggestion
            throw new Exception(sprintf(
                'You are trying to rename the column %s but you didn\'t change it\'s name',
                $oldName,
            ));
```

The `${foo}` syntax is deprecated and we don't allow string interpolation in this repository.

derrabus commented on src/Schema/TableDiff.php:
```diff
@@ -164,13 +167,12 @@ public function __construct(
         $removedForeignKeys = [],
-        $renamedColumns = [],
+        $renamedColumns = null,
         $renamedIndexes = []
```
Remove the parameter instead. The constructor is flagged as internal which means we own all calling code.

If you still want to build a deprecation layer, use `func_num_args()` and `func_get_arg()` to check for additional parameters.

derrabus commented on src/Schema/TableDiff.php:
```diff
@@ -134,6 +135,8 @@ class TableDiff
      */
     public $fromTable;
 
+    private bool $wasRenamedInternalUsed = false;
```
I don't think we need an additional property for that. Checking if `$renamedColumns` contains a non-empty value should be enough.

Tofandel replied:
The idea is to protect against external modification of the property directly without triggering a deprecation if the property was set internally, I can't do it without an additional property or removing the property and adding a `__set` method

All this deprecation layer should be removed in the next major, including this private

I can probably handle this in the constructor direclty a bit better though

derrabus commented on src/Schema/TableDiff.php:
```diff
@@ -238,10 +251,44 @@ public function getAddedColumns(): array
+        Deprecation::triggerIfCalledFromOutside(
+            'doctrine/dbal',
+            'https://github.com/doctrine/dbal/pull/6080',
+            '%s is deprecated, you should use `getModifiedColumns()` instead.',
```
```suggestion
            '%s is deprecated, use `getModifiedColumns()` instead.',
```

derrabus commented on src/Schema/TableDiff.php:
```diff
@@ -250,10 +297,30 @@ public function getDroppedColumns(): array
 
-    /** @return array<string,Column> */
+    /**
+     * @internal Use {@see getModifiedColumns()} instead.
```
```suggestion
     * @deprecated Use {@see getModifiedColumns()} instead.
```

Tofandel replied:
I put it internal because it's widely used in the tests and causes psalm errors if marked deprecated

```
ERROR: DeprecatedMethod - tests/Schema/AbstractComparatorTestCase.php:745:42 - The method Doctrine\DBAL\Schema\TableDiff::getRenamedColumns has been marked as deprecated (see https://psalm.dev/001)
        self::assertCount(0, $tableDiff->getRenamedColumns());
```

derrabus replied:
But that does not make sense. If you want that calling code should switch to a different method, that's by definition a deprecation.

And we are allowed to call our own deprecated methods. In fact, we need to do that sometimes in order to not lose coverage or break BC.

Have a look at psalm.xml. We have a huge list of deprecated symbols that we allow explicitly. You're free to exended it.

Tofandel replied:
I see, thanks. I added it there

I still think it can be a useful helper method though

Tofandel commented:
> I'm reviewing the PR now. I'm terribly sorry that I've postponed it for so long. One thing though: Codecov complains that we've lost coverage for the old API. Can you check that please?

No worries, we all have our lives to attend to as well ;)

Re the missing coverage, it's because I just moved around some lines including those old calls that don't have test coverage for the events returning false to prevent the rest of the execution, hence why the warnings are all on the `continue` line, but the events are still called correctly according to the tests, the events are already deprecated so I don't know if it's worth adding a test to have them return false for coverage

Re the deprecation layer I'll see what I can do

Tofandel commented on tests/Schema/TableDiffTest.php:
```diff
@@ -29,6 +33,59 @@ public function testReturnsName(): void
+
+        self::assertCount(2, $diff->renamedColumns);
+
+        $diff->renamedColumns = ['old_name' => new Column('new_name', Type::getType(Types::INTEGER))];
```
@derrabus This new abstraction layer is not perfect and contains a small breaking change, it's not possible to do `$diff->renamedColumns['old_name'] = new Column(..)` (throws an exception), it was marked as internal though and in laravel it's used like in the test case https://github.com/laravel/framework/blob/10.x/src/Illuminate/Database/Schema/Grammars/RenameColumn.php#L64-L67

So I think it's an acceptable internal breaking change

It is otherwise fixable but would involve the creation of a new Arrayable class to be returned from `__get` to call the setter when changes are pushed to the array

derrabus replied:
> So I think it's an acceptable internal breaking change

Totally.

> It is otherwise fixable but would involve the creation of a new Arrayable class to be returned from `__get` to call the setter when changes are pushed to the array

Let's not do down that road. 🙈

derrabus commented on src/Schema/Column.php:
```diff
@@ -463,4 +463,12 @@ public function toArray()
         ], $this->_platformOptions, $this->_customSchemaOptions);
     }
+
+    public function cloneWithName(string $name): Column
```
This method is only needed for the deprecation layer? I think, we should mark it as `@internal` then and remove it in 4.0.

greg0ire commented on src/Schema/Table.php:
```diff
@@ -346,6 +350,47 @@ public function addColumn($name, $typeName, array $options = [])
+
+        if ($oldName === $newName) {
+            throw new Exception(sprintf(
+                'You are trying to rename the column %s but you didn\'t change it\'s name',
```
```suggestion
                "You are trying to rename the column %s but you didn't change its name",
```

Also, looking at the project, we very rarely address the user directly. Also, you should probably throw a `LogicException` here as it is not meant to be caught.

greg0ire commented on src/Schema/TableDiff.php:
```diff
@@ -189,9 +179,50 @@ public function __construct(
+            Deprecation::trigger(
+                'doctrine/dbal',
+                'https://github.com/doctrine/dbal/pull/6080',
+                'Passing $renamedColumns to %s is deprecated and will be removed in the next major.',
```
"Passing" is not something that can be removed. Either say that it will no longer be possible, or that the argument will be removed.

greg0ire commented:
Your PR has 20 commits now. Are there any changes that could be extracted into smaller PRs and quickly merged? For instance, you mentioned [here](https://github.com/doctrine/dbal/pull/6080#discussion_r1246531475) that some tests were wrong.

greg0ire commented:
@Tofandel can you please try merging this up into 4.0.x? Or alternatively, send an equivalent PR to 4.0.x? @derrabus and I both tried the merge up, and it's a nightmare, mostly because of https://github.com/doctrine/dbal/pull/5788, which also touches a lot of the files you touched.

Tofandel commented:
@greg0ire When I have some time sure, I was looking at submitting it at 4.x originally because I saw how much code I was touching was already modified, but I was told to do it on 3.7. So yes I can imagine the nightmare, I think the main problem is that a big chunk of the tests that this rellies on where also modifed or removed

Tofandel commented:
@greg0ire I'm having a look, is it okay if I replace all arguments of TableDiff with named arguments? It's a class with a ton of them and it's hard to keep track of them, it will make this kind of changes easier to work with

![image](https://github.com/doctrine/dbal/assets/6115458/b5a0e8c2-2f25-4c63-b981-2ade82b6ccd3)

![image](https://github.com/doctrine/dbal/assets/6115458/9303ddb8-e49e-4970-9a5f-85f974e6da92)

greg0ire commented:
@Tofandel on 4.0.x, yes, on 3.8.x no, because named arguments were introduced in PHP 8.0

derrabus commented:
I've reverted the PR for now, in order to not block the new releases any further, see #6276. This buys us some time. Once we have a version of this feature for 4.x, we can restore your changes and do another 3.x feature release.

morozov commented on src/Schema/Table.php:
```diff
@@ -346,6 +351,48 @@ public function addColumn($name, $typeName, array $options = [])
+
+    /**
+     * @throws LogicException
+     * @throws SchemaException
```
@Tofandel where in this method is `SchemaException` thrown?

Tofandel replied:
It's not, i think it was a leftover because I was not sure which exception to chose and I started with SchemaException

morozov replied:
@Tofandel do you mind filing a PR that removes it? PhpStorm treats it as check exception meaning that the caller of this method should either catch it or have a `@throws SchemaException` annotation in its declaration. This is false-positive and should be fixed.
